nick_work/fdaaa_port.py

```diff
@@ -6,367 +6,310 @@
-           'intervention_mesh', 'keywords']
+#when actually implemented, point this towards the static location of the file in the cloud
+fda_reg_dict = {}
+with open("fdaaa_regulatory_snapshot.csv") as old_fda_reg:
+    reader = csv.DictReader(old_fda_reg)
+    for d in reader:
+        fda_reg_dict[d['nct_id']] = d['is_fda_regulated']
+
+
+def is_covered_phase(phase):
+    return phase in ['Phase 1/Phase 2', 'Phase 2', 'Phase 2/Phase 3' , 'Phase 3', 'Phase 4', 'N/A']
+
+def is_not_withdrawn(study_status):
+    return study_status != 'Withdrawn'
+
+def is_interventional(study_type):
+    return study_type == 'Interventional'
+
+def is_covered_intervention(intervention_type):
+    for it in intervention_type:
+        if ('Drug' or 'Device' or 'Biological' or 'Genetic' or
+           'Radiation' or 'Combination Prodcut' or 'Diagnostic Test') in it:
+            return True
+        else:
+            return False
+
+def is_not_device_feasibility(primary_purpose):
+    return primary_purpose != 'Device Feasibility'
+
+def is_fda_reg(fda_reg_drug, fda_reg_device):
+    if fda_reg_drug == 'Yes' or fda_reg_device == 'Yes':
+        return True
+    else:
+        return False
+
+def is_old_fda_regulated(is_fda_regulated, fda_reg_drug, fda_reg_device):
+    if ((fda_reg_drug is None and fda_reg_device is None) and
+        is_fda_regulated is not False):
+        return True
+    else:
+        return False
+    
+def has_us_loc(locs):
+    if ('United States' or 'American Samoa' or 'Guam' or 
+        'Northern Mariana Islands' or 'Puerto Rico' or 'Virgin Islands (U.S.)') in locs:
+        return True
+    else:
+        return False
+    
+def dict_or_none(data, keys):
+    for k in keys:
+        try:
+            data = data[k]
+        except KeyError:
+            return None
+    return json.dumps(data)
+
+#Some dates on clinicaltrials.gov are only Month-Year not Day-Month-Year. 
+#When this happens, we assign them to the last day of the month so our "results due" assessments are conservative
+def str_to_date(datestr):
+    is_defaulted_date = False
+    if datestr is not None:   
+        try:
+            parsed_date = datetime.strptime(datestr.text, '%B %d, %Y').date()
+        except ValueError:
+            parsed_date = datetime.strptime(datestr.text, '%B %Y').date() + relativedelta(months=+1) - timedelta(days=1)
+            is_defaulted_date = True
+    else:
+        parsed_date = None
+    return (parsed_date, is_defaulted_date)
+
+def t(textish):
+    if textish is None:
+        return None
+    return textish.text
+
+def does_it_exist(dataloc):
+    if dataloc is None:
+        return False
+    else:
+        return True
+
+def dict_builder(your_dict,name,value):
+    your_dict[name].append(value)
+    return your_dict
 
 effective_date = date(2017, 1, 18)
 directory = 'NCTxxx' #this points to a folder on my working directory. Change as needed.
 
-with open('test_output.csv', 'w') as test_csv:
+trial_dict = defaultdict(list)
```

Comment:
I want you to twist your table 90 degrees here.

Instead of making a dictionary containing lists, you should have made a list of dictionaries. Organize by a bunch of rows, not by a bunch of columns. (The danger of dict-of-lists is that you are later hoping that item \#574243 in list "foo" goes with the item \#574243 in list "bar", and that is super fragile. A later change where a loop over XML things doesn't insert a "foo" but does insert a "bar", and everything is wrong for the remaining ones.)

And in fact, I want to throw away the idea of holding many studies in memory at a time. There is no need to. Construct one, then write it out and move on.

So, when you read in a XML file and parse it, make a new, single dictionary there. Later, hand that dictionary to the CSV writer before you go to the next study.

Catch all values in the dict, instead of local variables.

```
    s["phase"] = t(soup.phase)
```

Reply:
This makes sense and definitely easier to implement! The one thing to note is that some variables both go into the resulting CSV and are used again within the expressions or functions that determine other variables so I think in those instances, I'll have to both set the variable and assign it to the dictionary in 2 separate lines.

Reply:
Er, maybe I don't understand, but dictionaries aren't write-only. You can use `s['thingpreviouslyset"]` to read as well as write. I guess you might be mutating it into something else you don't want also to store, so in some cases you might want an extra variable.